internal/controllers/topology/cluster/structuredmerge/dryrun.go

```diff
@@ -0,0 +1,301 @@
+	if modifiedList, ok := ctx.modified.([]interface{}); ok {
+		// NOTE: ignoring the error in case the element wasn't in original.
+		originalList, _ := ctx.original.([]interface{})
+
+		// Process listType = map/set, previously not empty.
+		// NOTE: listType = map/set but previously empty is managed like scalar values down in the func.
+		if len(ctx.fieldsV1) != 0 {
+			// If the number of items is changed from the previous reconcile it is already clear that
+			// something is changed without checking all the items.
+			// NOTE: this assumes the root of the object isn't a list, which is true for all the K8s objects.
+			if len(modifiedList) != len(ctx.fieldsV1) || len(modifiedList) != len(originalList) {
+				return pathToResult(ctx.path)
+			}
+
+			// Otherwise, check the item in the list one by one.
+
+			// if the list is a listMap
+			if isListMap(ctx.fieldsV1) {
+				for itemKeys, itemFieldsV1 := range ctx.fieldsV1 {
+					// Get the keys for the current item.
+					keys := getFieldV1Keys(itemKeys)
+
+					// Get the corresponding original and modified item.
+					modifiedItem := getItemWithKeys(modifiedList, keys)
+					originalItem := getItemWithKeys(originalList, keys)
+
+					// Get the managed field for this item.
+					// NOTE: ignoring conversion failures because itemFieldsV1 are always of this type.
+					fieldV1Map, _ := itemFieldsV1.(map[string]interface{})
+
+					// Check for changes in the item value.
+					itemHasChanges, itemHasSpecChanges := dryRunPatch(&dryRunInput{
+						path:     ctx.path,
+						fieldsV1: fieldV1Map,
+						modified: modifiedItem,
+						original: originalItem,
+					})
+					hasChanges = hasChanges || itemHasChanges
+					hasSpecChanges = hasSpecChanges || itemHasSpecChanges
+
+					// If there are already changes detected, it is possible to skip processing other items.
+					if hasChanges {
+						break
+					}
+				}
+				return
+			}
+
+			if isListSet(ctx.fieldsV1) {
+				s := sets.NewString()
+				for v := range ctx.fieldsV1 {
+					s.Insert(strings.TrimPrefix(v, "v:"))
+				}
+
+				for _, v := range modifiedList {
+					// NOTE: ignoring this error because API server ensures the keys in listMap are scalars value.
+					vString, _ := v.(string)
+					if !s.Has(vString) {
+						return pathToResult(ctx.path)
+					}
+				}
+				return
+			}
+		}
+
+		// NOTE: listType = atomic is managed like scalar values down in the func;
+		// a list is atomic when the corresponding FieldV1 doesn't have info for the list items.
+	}
+
+	// Otherwise, the func is processing scalar or atomic values.
+
+	// Check if the field has been added (it wasn't managed before).
+	// NOTE: This prevents false positive when handling metadata, because it is required to have metadata.name and metadata.namespace
+	// in modified, but they are not tracked as managed field.
+	notManagedBefore := ctx.fieldsV1 == nil
+	if len(ctx.path) == 1 && ctx.path[0] == "metadata" {
```

Comment:
I'm not sure I understand this case. Essentially we can't conclude that metadata is unmanaged just because the top-level metadata doesn't have fieldsV1?

And this case occurs if in metadata only name/space are set in modified? (and managed fields for them is empty then?)

Reply:
If I get it right: this essentialy says that `metadata` was always managedBefore (== `notManagedBefore=false`) because `name` and `namespace` should always be considered as managed by "me" / "whichever manager you are" / --> all managers which act on the obj (and due to immutability of the fields this is totally fine).

Maybe we could rephrase:
```
	// NOTE: This prevents false positive when handling metadata. The fields metadata.name and metadata.namespace are 
	// always implicitly managed before, but this fields are not tracked as managed field.
```

wdyt? (cc @fabriziopandini )

Reply:
This is correct. 
Metadata was always managedBefore (== notManagedBefore=false) because name and namespace should always be considered as managed

Reply:
I think it mostly confuses my because it's coupled to the notManagedBefore variable.
But I guess the situation is the following, if:
* we are currently in metadata (ctx.path == "metadata")
* and fieldsV1 are nil

We can't just assume that something changed (which is what setting notManagedBefore to true leads to eventually).

And that is because if only metadata.name and metadata.namespace are set fieldsV1 will be nil.